### PR TITLE
Data Management API: Created TransferProcessService

### PR DESCRIPTION
**Pull request by algattik**
## What this PR changes/adds

Add a TransferProcessService, a service in the Data Management API to manage transfer processes.

#809 implementation was split into this PR and PR #1006.

## Why it does that

Data Management API surface

## Further notes

Moved `Command` objects to the SPI package where they belong. This avoids having to create a dependency from an extension package to core package.

**Note that Deprovisioning command is enqueued also if the TP is in ENDED or CANCELLED states, given this is a permissible transition per existing code.**

## Linked Issue(s)

Relates to #809  

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [x] documented public classes/methods?
- [x] added/updated relevant documentation?
- [x] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: extensions/api/data-management/transferprocess/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/transferprocess/service/TransferProcessServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    @Override
+    public @NotNull Collection<TransferProcess> query(QuerySpec query) {
+        return transactionContext.execute(() -> transferProcessStore.findAll(query).collect(toList()));
+    }
+
+    @Override
+    public @Nullable String getState(String transferProcessId) {
+        return transactionContext.execute(() -> {
+            var process = transferProcessStore.find(transferProcessId);
+            if (process == null) {
```
[the remark this answers is not in the document]
Comment (algattik) [reply extensions/api/data-management/transferprocess/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/transferprocess/service/TransferProcessServiceImpl.java:57]: Yes Optional would be very useful in many places, but my understanding is it's not used in the entire codebase. @paullatzelsperger @jimmarino any feedback on that?

File: extensions/api/data-management/transferprocess/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/transferprocess/service/TransferProcessServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+    @Test
+    void getState_whenNotFound() {
+        assertThat(service.getState(id)).isNull();
+    }
+
+    @ParameterizedTest
+    @MethodSource("cancellableStates")
+    void cancel(TransferProcessStates state) {
+        // Arrange
```
[the remark this answers is not in the document]
Comment (algattik) [reply extensions/api/data-management/transferprocess/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/transferprocess/service/TransferProcessServiceImplTest.java:106]: For tests that are more than a few lines long I find them useful for maintenance, they're also used extensively in the codebase. @paullatzelsperger any opinion?

File: extensions/api/data-management/transferprocess/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/transferprocess/service/TransferProcessServiceImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ndr-brt) [reply extensions/api/data-management/transferprocess/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/transferprocess/service/TransferProcessServiceImpl.java:57]: As far as I remember we decided to not use `Optional` for method return types, but they are used in a lot of places to handle the "do stuff unless null" pattern like this

File: extensions/api/data-management/transferprocess/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/transferprocess/service/TransferProcessServiceImplTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ndr-brt) [reply extensions/api/data-management/transferprocess/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/transferprocess/service/TransferProcessServiceImplTest.java:106]: I think their "extensive use" was caused by a lot of copy-and-paste between test, a correct empty lines usage to separate arrange-act-assert already does the job nicely.

File: extensions/api/data-management/transferprocess/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/transferprocess/service/TransferProcessServiceImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (algattik) [reply extensions/api/data-management/transferprocess/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/transferprocess/service/TransferProcessServiceImpl.java:57]: ah you mean not to change the signature of `find`, but just to use `Optional.ofNullable(...).map(...)` for routing this code? Functional vs imperative style is a bit of a matter of preference

[the remark this answers is not in the document]
Comment (algattik) [reply extensions/api/data-management/transferprocess/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/transferprocess/service/TransferProcessServiceImpl.java]: Implemented

File: extensions/api/data-management/transferprocess/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/transferprocess/service/TransferProcessService.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    /**
+     * Asynchronously requests cancellation of the transfer process.
+     * <p>
+     * The return result status only reflects the successful submission of the command.
+     *
+     * @param transferProcessId id of the transferProcess
+     * @return a result that is successful if the transfer process was found and is in a state that can be canceled
+     */
+    @NotNull
+    Result<?> cancel(String transferProcessId);
```
[the remark this answers is not in the document]
Comment (algattik) [reply extensions/api/data-management/transferprocess/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/transferprocess/service/TransferProcessService.java]: Implemented

File: extensions/api/data-management/transferprocess/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/transferprocess/service/TransferProcessServiceImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ndr-brt) [reply extensions/api/data-management/transferprocess/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/transferprocess/service/TransferProcessServiceImpl.java]: yes, but I feel like in these cases functional manner works better, and it's used also by `TransferProcessStates.from`, that's used here, so it's pretty coherent

**Review by ndr-brt — APPROVED** [review]
let's get this merged

**Review by paullatzelsperger — APPROVED** [review]

**Comment by jimmarino** [comment]
@algattik Changelog and I'll merge

**Pull request merged**